lib/AST/IRToASTVisitor.cpp

```diff
@@ -61,9 +63,86 @@ class ExprGen : public llvm::InstVisitor<ExprGen, clang::Expr *> {
   clang::Expr *visitUnaryOperator(llvm::UnaryOperator &inst);
 };
 
+clang::Expr *IRToASTVisitor::ConvertExpr(z3::expr expr) {
+  if (expr.decl().decl_kind() == Z3_OP_EQ) {
+    CHECK_EQ(expr.num_args(), 2) << "Equalities must have 2 arguments";
+    auto a{expr.arg(0)};
+    auto b{expr.arg(1)};
+
+    llvm::SwitchInst *inst{dec_ctx.z3_sw_vars_inv[a.id()]};
+    unsigned case_idx{};
+
+    if (!inst) {
+      inst = dec_ctx.z3_sw_vars_inv[b.id()];
+      case_idx = a.get_numeral_uint();
+    } else {
+      case_idx = b.get_numeral_uint();
+    }
+
+    for (auto sw_case : inst->cases()) {
+      if (sw_case.getCaseIndex() == case_idx) {
+        return ast.CreateEQ(CreateOperandExpr(inst->getOperandUse(0)),
+                            CreateConstantExpr(sw_case.getCaseValue()));
+      }
+    }
+
+    LOG(FATAL) << "Couldn't find switch case";
+  }
+
+  auto hash{expr.id()};
+  if (dec_ctx.z3_br_edges_inv.find(hash) != dec_ctx.z3_br_edges_inv.end()) {
+    auto edge{dec_ctx.z3_br_edges_inv[hash]};
+    CHECK(edge.second) << "Inverse map should only be populated for branches "
+                          "taken when condition is true";
+    return CreateOperandExpr(*(edge.first->op_end() - 3));
```

Comment:
Does `*(edge.first->op_end() - 3)` return the branch condition? Definitely needs a comment or a more idiomatic way of getting the operand :eyes:

Reply:
I couldn't find public API for what I needed, so I added a comment